File: app/src/main/java/com/coinninja/coinkeeper/cn/account/RemoteAddressCache.java
package com.coinninja.coinkeeper.cn.account;

import com.coinninja.coinkeeper.cn.wallet.HDWallet;
import com.coinninja.coinkeeper.model.dto.AddressDTO;
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient;
import com.coinninja.coinkeeper.service.client.model.CNWalletAddress;
import com.coinninja.coinkeeper.util.RemoteAddressLocalCache;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

import retrofit2.Response;

@SuppressWarnings("ALL")
public class RemoteAddressCache {
    private static final int NUM_ADDRESSES_TO_CACHE = 5;
    private final SignedCoinKeeperApiClient apiClient;
    private final AccountManager accountManager;
    private final RemoteAddressLocalCache remoteAddressLocalCache;

    @Inject
    RemoteAddressCache(SignedCoinKeeperApiClient apiClient, AccountManager accountManager, RemoteAddressLocalCache remoteAddressLocalCache) {
        this.apiClient = apiClient;
        this.accountManager = accountManager;
        this.remoteAddressLocalCache = remoteAddressLocalCache;
    }

    public void cacheAddresses() {
        Response response = apiClient.getCnWalletAddresses();

        if (response.isSuccessful()) {
            List<CNWalletAddress> cachedAddresses = (List<CNWalletAddress>) response.body();

            HashMap<String, AddressDTO> addressStringToDTO = accountManager
                    .unusedAddressesToPubKey(HDWallet.EXTERNAL, NUM_ADDRESSES_TO_CACHE);

            removeUsedAddresses(cachedAddresses, addressStringToDTO);
            removeServerAddressesWithoutPublicKey(cachedAddresses);
            sendAddresses(cachedAddresses, addressStringToDTO);
        }

    }

    private void cacheUnusedAddressesLocally(List<AddressDTO> unusedAddresses) {
        remoteAddressLocalCache.setLocalRemoteAddressCache(unusedAddresses);
    }

    private void removeServerAddressesWithoutPublicKey(List<CNWalletAddress> cachedAddresses) {
        for (CNWalletAddress address : cachedAddresses) {
            if (address.getPublicKey() == null || address.getPublicKey().isEmpty()) {
                apiClient.removeAddress(address.getAddress());
            }
        }
    }

    private void removeUsedAddresses(List<CNWalletAddress> cachedAddresses, HashMap<String, AddressDTO> idToUnusedAddress) {
        for (CNWalletAddress walletAddress : cachedAddresses) {
            if (idToUnusedAddress.get(walletAddress.getAddress()) == null)
                apiClient.removeAddress(walletAddress.getAddress());
        }
    }

    private void sendAddresses(List<CNWalletAddress> cachedAddresses, HashMap<String, AddressDTO> addressDTOHashMap) {
        HashMap<String, AddressDTO> addressToDTOMapCopy = new HashMap(addressDTOHashMap);

        for (CNWalletAddress walletAddress : cachedAddresses) {
            if (addressDTOHashMap.get(walletAddress.getAddress()) != null) {
                addressToDTOMapCopy.remove(walletAddress.getAddress());
            }
        }

        for (String address : addressToDTOMapCopy.keySet()) {
            AddressDTO addressToSend = addressToDTOMapCopy.get(address);
            apiClient.addAddress(address, addressToSend.getUncompressedPublicKey());
        }

        cacheUnusedAddressesLocally(new ArrayList<>(addressDTOHashMap.values()));
    }

}

File: app/src/main/java/com/coinninja/coinkeeper/cn/account/RemoteAddressCache.kt
package com.coinninja.coinkeeper.cn.account

import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.cn.wallet.HDWalletWrapper
import com.coinninja.coinkeeper.model.dto.AddressDTO
import com.coinninja.coinkeeper.model.helpers.WalletHelper
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import com.coinninja.coinkeeper.service.client.model.AddAddressBodyRequest
import com.coinninja.coinkeeper.service.client.model.CNWalletAddress
import com.coinninja.coinkeeper.util.RemoteAddressLocalCache
import retrofit2.Response
import java.util.*
import javax.inject.Inject

@Mockable
@Suppress("UNCHECKED_CAST")
class RemoteAddressCache @Inject internal constructor(
        internal val hdWalletWrapper: HDWalletWrapper,
        internal val apiClient: SignedCoinKeeperApiClient,
        internal val accountManager: AccountManager,
        internal val walletHelper: WalletHelper,
        internal val remoteAddressLocalCache: RemoteAddressLocalCache
) {
    fun cacheAddresses() {
        val response: Response<*> = apiClient.cnWalletAddresses
        if (response.isSuccessful) {
            val cachedAddresses: List<CNWalletAddress> = response.body() as List<CNWalletAddress>
            val addressStringToDTO: HashMap<String, AddressDTO> = accountManager
                    .unusedAddressesToPubKey(HDWalletWrapper.EXTERNAL, NUM_ADDRESSES_TO_CACHE)
            removeUsedAddresses(cachedAddresses, addressStringToDTO)
            removeServerAddressesWithoutPublicKey(cachedAddresses)
            sendAddresses(cachedAddresses, addressStringToDTO)
            addLightning(cachedAddresses)
        }
    }

    private fun addLightning(cachedAddresses: List<CNWalletAddress>) {
        var hasLightning = false
        walletHelper.segwitWallet?.let { wallet ->
            cachedAddresses.forEach { walletAddress ->
                if (walletAddress.address == "generate" && walletAddress.publicKey == hdWalletWrapper.verificationKeyFor(wallet)) {
                    hasLightning = true
                } else if (walletAddress.address == "generate") {
                    apiClient.removeAddress("generate")
                }
            }

            if (!hasLightning && wallet.purpose == 84) {
                apiClient.addAddress(AddAddressBodyRequest(pubKey = hdWalletWrapper.verificationKeyFor(wallet), addressType = "lightning"))
            }
        }

    }

    private fun cacheUnusedAddressesLocally(unusedAddresses: List<AddressDTO>) {
        remoteAddressLocalCache.localRemoteAddressCache = unusedAddresses
    }

    private fun removeServerAddressesWithoutPublicKey(cachedAddresses: List<CNWalletAddress>) {
        for (address in cachedAddresses) {
            if (address.publicKey.isNullOrEmpty()) {
                apiClient.removeAddress(address.address)
            }
        }
    }

    private fun removeUsedAddresses(cachedAddresses: List<CNWalletAddress>, localCache: HashMap<String, AddressDTO>) {
        for (walletAddress in cachedAddresses) {
            if (walletAddress.address == "generate") continue
            if (!localCache.keys.contains(walletAddress.address))
                apiClient.removeAddress(walletAddress.address)
        }
    }

    private fun sendAddresses(cachedAddresses: List<CNWalletAddress>, addressDTOHashMap: HashMap<String, AddressDTO>) {
        val cachedAddrs: MutableList<String> = mutableListOf()

        cachedAddresses.forEach { walletAddress ->
            cachedAddrs.add(walletAddress.address)
        }

        addressDTOHashMap.values.forEach { dto ->
            if (!cachedAddrs.contains(dto.address))
                apiClient.addAddress(AddAddressBodyRequest(address = dto.address, pubKey = dto.uncompressedPublicKey))
        }

        cacheUnusedAddressesLocally(ArrayList(addressDTOHashMap.values))
    }

    fun removeAll() {
        apiClient.cnWalletAddresses.let { response ->
            if (response.isSuccessful) {
                response.body()?.let { remoteAddresses ->
                    remoteAddresses.forEach {
                        apiClient.removeAddress(it.address)
                    }
                }

                remoteAddressLocalCache.localRemoteAddressCache = emptyList()
            }
        }
    }

    companion object {
        private const val NUM_ADDRESSES_TO_CACHE = 5
    }

}